### PR TITLE
Add db thread pool option to keep idle connections alive

### LINKED ISSUE #4675: "OperationalError: server closed the connection unexpectedly" on SQL: ‘SELECT … FROM organizations’
From time to time the UI start getting “internal server error” after a couple of refreshes it is start working again normally. This happens on every screen: login, dashboard, queries, etc… I can see in the log that redash-server is trying to make the following SQL query:

`[SQL: 'SELECT organizations.updated_at AS organizations_updated_at, organizations.created_at AS organizations_created_at, organizations.id AS organizations_id, organizations.name AS organizations_name, organizations.slug AS organizations_slug, organizations.settings AS organizations_settings \nFROM organizations \nWHERE organizations.slug = %(slug_1)s \n LIMIT %(param_1)s'] [parameters: {'slug_1': 'default', 'param_1': 1}] (Background on this error at: http://sqlalche.me/e/e3q8)`

But it fails with **OperationalError: (psycopg2.OperationalError) server closed the connection unexpectedly**

I have selfhosted redash with the following setup:

standalone postgre machine
standalone redis machine
docker redash-server: redash/redash:8.0.2.b37747
docker redash-scheduler: redash/redash:8.0.2.b37747
docker redash-worker: redash/redash:8.0.2.b37747
docker ad-hoc-worker: redash/redash:8.0.2.b37747
The error stacktrace is:

```
Feb 22, 2020 @ 10:19:44.908	    cursor.execute(statement, parameters)
Feb 22, 2020 @ 10:19:44.908	  File "/usr/local/lib/python2.7/site-packages/flask_login/login_manager.py", line 317, in _load_user
Feb 22, 2020 @ 10:19:44.908		This probably means the server terminated abnormally
Feb 22, 2020 @ 10:19:44.908		before or while processing the request.
Feb 22, 2020 @ 10:19:44.908	 [SQL: 'SELECT organizations.updated_at AS organizations_updated_at, organizations.created_at AS organizations_created_at, organizations.id AS organizations_id, organizations.name AS organizations_name, organizations.slug AS organizations_slug, organizations.settings AS organizations_settings \nFROM organizations \nWHERE organizations.slug = %(slug_1)s \n LIMIT %(param_1)s'] [parameters: {'slug_1': 'default', 'param_1': 1}] (Background on this error at: http://sqlalche.me/e/e3q8)
Feb 22, 2020 @ 10:19:44.908	    response = self.full_dispatch_request()
Feb 22, 2020 @ 10:19:44.908	  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1413, in _handle_dbapi_exception
Feb 22, 2020 @ 10:19:44.908	    exc_info
Feb 22, 2020 @ 10:19:44.908	  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/util/compat.py", line 265, in raise_from_cause
Feb 22, 2020 @ 10:19:44.908	    context)
Feb 22, 2020 @ 10:19:44.908	  File "/usr/local/lib/python2.7/site-packages/flask/app.py", line 1982, in wsgi_app
Feb 22, 2020 @ 10:19:44.908	    reraise(type(exception), exception, tb=exc_tb, cause=cause)
Feb 22, 2020 @ 10:19:44.908	  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1193, in _execute_context
Feb 22, 2020 @ 10:19:44.908	  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/engine/default.py", line 509, in do_execute
Feb 22, 2020 @ 10:19:44.908	OperationalError: (psycopg2.OperationalError) server closed the connection unexpectedly
Feb 22, 2020 @ 10:19:44.907	Traceback (most recent call last):
Feb 22, 2020 @ 10:19:44.907	  File "/usr/local/lib/python2.7/site-packages/flask/app.py", line 1609, in full_dispatch_request
Feb 22, 2020 @ 10:19:44.907	    user = self.user_callback(user_id)
Feb 22, 2020 @ 10:19:44.907	  File "/usr/local/lib/python2.7/site-packages/flask/app.py", line 1614, in full_dispatch_request
Feb 22, 2020 @ 10:19:44.907	    request_started.send(self)
Feb 22, 2020 @ 10:19:44.907	  File "/app/redash/authentication/__init__.py", line 48, in load_user
Feb 22, 2020 @ 10:19:44.907	    org = current_org._get_current_object()
Feb 22, 2020 @ 10:19:44.907	  File "/usr/local/lib/python2.7/site-packages/werkzeug/local.py", line 302, in _get_current_object
Feb 22, 2020 @ 10:19:44.907	    return self.__local()
Feb 22, 2020 @ 10:19:44.907	  File "/app/redash/authentication/org_resolving.py", line 18, in _get_current_org
Feb 22, 2020 @ 10:19:44.907	    g.org = Organization.get_by_slug(slug)
Feb 22, 2020 @ 10:19:44.907	  File "/app/redash/models/organizations.py", line 33, in get_by_slug
Feb 22, 2020 @ 10:19:44.907	    return cls.query.filter(cls.slug == slug).first()
Feb 22, 2020 @ 10:19:44.907	  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2895, in first
Feb 22, 2020 @ 10:19:44.907	    ret = list(self[0:1])
Feb 22, 2020 @ 10:19:44.907	  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2687, in __getitem__
Feb 22, 2020 @ 10:19:44.907	    return list(res)
Feb 22, 2020 @ 10:19:44.907	  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2995, in __iter__
Feb 22, 2020 @ 10:19:44.907	    return self._execute_and_instances(context)
Feb 22, 2020 @ 10:19:44.907	  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 3018, in _execute_and_instances
Feb 22, 2020 @ 10:19:44.907	    result = conn.execute(querycontext.statement, self._params)
Feb 22, 2020 @ 10:19:44.907	  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 948, in execute
Feb 22, 2020 @ 10:19:44.907	    return meth(self, multiparams, params)
Feb 22, 2020 @ 10:19:44.907	  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/sql/elements.py", line 269, in _execute_on_connection
Feb 22, 2020 @ 10:19:44.907	    return connection._execute_clauseelement(self, multiparams, params)
Feb 22, 2020 @ 10:19:44.907	  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1060, in _execute_clauseelement
Feb 22, 2020 @ 10:19:44.907	    compiled_sql, distilled_params
Feb 22, 2020 @ 10:19:44.907	  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1200, in _execute_context
Feb 22, 2020 @ 10:19:44.907	    context)
Feb 22, 2020 @ 10:19:44.907	    rv = self.handle_user_exception(e)
Feb 22, 2020 @ 10:19:44.907	  File "/usr/local/lib/python2.7/site-packages/blinker/base.py", line 267, in send
Feb 22, 2020 @ 10:19:44.907	    for receiver in self.receivers_for(sender)]
Feb 22, 2020 @ 10:19:44.907	  File "/app/redash/models/users.py", line 54, in update_user_active_at
Feb 22, 2020 @ 10:19:44.907	    if current_user.is_authenticated and not current_user.is_api_user():
Feb 22, 2020 @ 10:19:44.907	  File "/usr/local/lib/python2.7/site-packages/werkzeug/local.py", line 343, in __getattr__
Feb 22, 2020 @ 10:19:44.907	    return getattr(self._get_current_object(), name)
Feb 22, 2020 @ 10:19:44.907	  File "/usr/local/lib/python2.7/site-packages/werkzeug/local.py", line 302, in _get_current_object
Feb 22, 2020 @ 10:19:44.907	    return self.__local()
Feb 22, 2020 @ 10:19:44.907	  File "/usr/local/lib/python2.7/site-packages/flask_login/utils.py", line 26, in <lambda>
Feb 22, 2020 @ 10:19:44.907	    current_user = LocalProxy(lambda: _get_user())
Feb 22, 2020 @ 10:19:44.907	  File "/usr/local/lib/python2.7/site-packages/flask_login/utils.py", line 302, in _get_user
Feb 22, 2020 @ 10:19:44.907	    current_app.login_manager._load_user()
Feb 22, 2020 @ 10:19:44.907	  File "/usr/local/lib/python2.7/site-packages/flask/app.py", line 1517, in handle_user_exception
Feb 22, 2020 @ 10:19:44.907	    return self.reload_user()
Feb 22, 2020 @ 10:19:44.907	  File "/usr/local/lib/python2.7/site-packages/flask_restful/__init__.py", line 271, in error_router
Feb 22, 2020 @ 10:19:44.907	    return original_handler(e)
Feb 22, 2020 @ 10:19:44.907	[2020-02-22 08:19:44,905] ERROR in app: Exception on /api/organization/status [GET]
Feb 22, 2020 @ 10:19:44.907	    reraise(exc_type, exc_value, tb)
Feb 22, 2020 @ 10:19:44.907	  File "/usr/local/lib/python2.7/site-packages/flask_login/login_manager.py", line 279, in reload_user
```

However thank you for this great tool ! Appreciate your work :)

### PR DESCRIPTION
**Pull request by gstaykov**
This PR is fixing the following bug with idle db connections. 

Closes #4675.

### CODE REVIEW

File: redash/models/base.py [hunk 1]
```diff
@@ -12,10 +12,12 @@
 class RedashSQLAlchemy(SQLAlchemy):
     def apply_driver_hacks(self, app, info, options):
         options.update(json_serializer=json_dumps)
+        options.update(pool_pre_ping=True)
```
Comment (arikfr) [inline redash/models/base.py]: Are you sure it's needed here too?

Comment (gstaykov) [reply redash/models/base.py]: Yes. Without it the bug is still there. I have tested it.

I will update the pull request with the recommendations later :)

File: redash/settings/__init__.py [hunk 2]
```diff
@@ -36,6 +36,9 @@
 SQLALCHEMY_DISABLE_POOL = parse_boolean(
     os.environ.get("SQLALCHEMY_DISABLE_POOL", "false")
 )
+SQLALCHEMY_ENABLE_POOL_PRE_PING = parse_boolean(
+   os.environ.get("SQLALCHEMY_ENABLE_POOL_PRE_PING", "true")
```
Comment (arikfr) [inline redash/settings/__init__.py:40]:
```suggestion
   os.environ.get("SQLALCHEMY_ENABLE_POOL_PRE_PING", "false")
```

Default value should be false to keep current behavior.

**Comment by gstaykov** [comment]
If you set it by default false it is like saying "ok your UI will throw exception after every 30 minutes of inactivity. So when you come back to the UI after 30 minutes you need to hit refresh 5-6 times to start working again with the UI" ....    In my opinion it should be true by default because this is major bug I can't see a case where someone will set this to false ....

**Comment by arikfr** [comment]
I've been running Redash in multiple production environments for the last 6+ years and never had this issue.

I accept that in some configurations this can happen and therefore it's fine to have this option, but I don't want to impact performance of most deployments just because some have this issue.

**Comment by gstaykov** [comment]
Ok you can test it with the following steps:

1. Open redash UI
2. Don't use it for at least 30 minutes. Everyone need to stop using it. There must be no requests to the backend for 30 minutes (wait the db connections from the pool to die).
3. Try to use it again.

I'm pretty sure everyone have this problem. Except if they don't use some other way to keep their db connections alive (use the UI all day without stops, make/schedule requests to the backend every minute, etc...)

**Comment by arikfr** [comment]
@gstaykov or you can try the following:

1. Deploy Redash with our default setup (https://redash.io/help/open-source/setup).
2. Do your experiment.

I checked logs on the [redash-preview](https://redash-preview.netlify.com/) instance and I can see that there are times that requests come after >30 minutes without an error:

```
server_1            | [2020-04-03 02:17:45,867][PID:58][INFO][metrics] method=GET path=/api/session endpoint=redash_session status=404 content_type=application/json content_length=66 duration=0.33 query_count=0 query_duration=0.00
server_1            | [2020-04-03 03:00:52,631][PID:62][INFO][metrics] method=GET path=/ endpoint=redash_index status=302 content_type=text/html; charset=utf-8 content_length=293 duration=0.59 query_count=0 query_duration=0.00
```

**Comment by gstaykov** [comment]
I'm trying to integrate redash in my company and we have fresh install of the latest version and we hit this issue every day in the morning when we try to login. I describe it in details here: https://github.com/getredash/redash/issues/4675   Also I saw that other people are hitting this bug too (from the https://discuss.redash.io/search?q=server%20closed%20the%20connection%20unexpectedly)   However I like you product very much thanks for your work :)

**Comment by gstaykov** [comment]
This is the error "OperationalError: (psycopg2.OperationalError) server closed the connection unexpectedly" You can search it in logs and forum to see how often people are hitting it :)

**Comment by arikfr** [comment]
> You can search it in logs 

I didn't find it in the logs.

> and forum to see how often people are hitting it :)

Based on the forum, there were 3-5 people who had this error. Not that many. It will be great to provide them this option, but no reason to punish the thousands of others who use Redash and don't get this error.

Also, in your case you're using a standalone Postgres. Can you provide some details about it: what version it runs? How was it setup (is it something like RDS)?

**Comment by arikfr** [comment]
(just noticed the update to your comment)

> This redash-preview is not a good example because many people are using it and there are many requests to the backend.

I looked at the logs and found a period of time where there was >30 minutes between two requests. Between these two requests, the server didn't use any of its connections.

You're welcome to do the experiment with a test environment yourself and report back. I'm pretty sure the difference here is how you configure your Postgres database.

**Comment by gstaykov** [comment]
We have our own Postgres on our own machine. The version is PostgreSQL 9.6.17 on x86_64-pc-linux-gnu, compiled by gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-39), 64-bit.  The setup is the default using the official documentation and nothing special.

I think you make the right decision. If many people are using this option in future (i guess there is a log) then you can make it "true" by default 👍

**Comment by rcoup** [comment]
FWIW, we're hitting it with Redshift as a backend reasonably regularly.

**Pull request merged**